### ISSUE TITLE
[Feature Pitch] Add replay system to KAG

### ISSUE THREAD

**Issue by deynarde** [transhumandesign/kag-base#216]
A replay system would record all players’ actions, save it to a file and enable replaying it later.

**Why implement it?**

- Replay system can be used to show off your best moments in KAG without recording a video, for example you could save a good example of teamwork between a knight and a builder or a 5+ kills keg drop.
- It can be used to remake existing KAG tutorial and help new players learning the game by showing them real examples of common situations (spiking a group of enemy players, how to counter the enemy builder etc.)
- Can help with catching griefers and rulebreakers - you could save a replay of a griefer, then give it to an admin.
- You could save interesting CTF matches that you would want to watch later.
- Can help with making more interesting videos/montages from different perspectives (enemy’s perspective, or perhaps an item’s perspective - for example, a bomb’s perspective). Would give more possibilities to video makers like The_Osz.   
- It is much more lightweight than a video.

**How would it work?**

Eluded made a replay mod back in 2016, where you could use various commands like `!record`, `!stoprecording`, `!replay` to record character positions and replay them later. My idea is to expand upon this mod and enable players to record whole CTF matches, although you also could choose to record only a few minutes.

Here’s a GIF how the mod works:
https://gfycat.com/CookedThirdLeafcutterant

Recordings should be saved to a file, which you could play in either multiplayer or solo. You could then cut it at certain parts, speed something up, slow something down, change perspectives [spectator perspective, blue knight perspective, red archer perspective, maybe even a bomb’s perspective]. 

To make things easier, there probably should be a time bar like this [screenshot from Rocket League]:
![image](https://user-images.githubusercontent.com/10031276/44981759-da196f80-af73-11e8-82c8-a69d67753f3b.png)

You could see which minute of a replay you are on, you could also see the most important events and their timeframes [like a flag pick up, flag capture, next map, etc.].

An additional idea for future would be a replay sharing system; you could upload your replay to some community hub. You could upvote other replays, and we could have some categories like ‘best replays of the week’, ‘best replays of the month’, ‘new replays’, etc.

Eluded's mod link: 
https://github.com/benjamingorman/KAGActionReplay

**Comment by Suprejso** [comment]
I thought of it a few times. Definitely one of the best ideas out there at the moment. Though it would take some time to be made, if you are talking about saving whole matches as .replay files, as it probably needs a new "Replay" button in the menu. The thing is that KAG is already laggy on weak devices, replay system would make it even more laggy for them.
Anyways, I'm really looking forward for it to be made, totally appreciate the idea.

**Comment by Blubahub** [comment]
Yes - I need this. Maybe also write it in such a way that it would work with mods (e.g. make it a weird .replay interactive but recorded video file for KAG only)? Also, it would be nice if you could export it as a e.g. mp4, .mov, and any other video file types/extensions (not saying to do them all - just choose one or two).

**Comment by Mazey** [comment]
Not a feature pitch. Pretty much of this will have to be done engine-side (it *would* be possible to do with scripts but not feasible at all).

**Comment by deynarde** [comment]
If this feature pitch gets popular enough i think Asu and Geti could help with some engine related changes as they did in **Feature Pitch: Improve failure rate of knight animations**.

> Some of this is probably beyond the reach of non-engine contributions but should this turn out to be the most-requested change we'll happily work with contributors on the bounty and investigate any engine changes that could help alleviate the issue.

And as you can see in Eluded's mod it's possible to get a working prototype done with non engine changes.

**Comment by AsuMagic** [comment]
Implementing something like that is obviously not trivial.

At first I thought of recording net commands in-engine, which would be extremely fast (as all there is to do is to store the received buffers) but brings a lot of complications for replaying. If the engine was written with that in mind it would surely be okay, but it isn't afaict.

The other solution is to keep track of properties you want to look for, which is simpler in-engine because things like `CAnimSprite` (internal dependency of `CSprite`) are optimized internally so we can store state fairly easily.

But depending on how this is implemented, this has the problem of not being able to record what a client doesn't see, when it might be desirable to record what's happening all over the map... which means doing it server-side.

I can see a nice usecase to this too: something like Overwatch's POTG, which uses a heuristic to determine the best action in a game and replays it at the end of a match.  
It would be nice for CTF/TTH, but might be undesirable for TDM.

In general, though, the performance of such a system implemented in scripts might be not desirable.

It would be especially ideal if players could export a record to h.264 and upload directly to youtube, but that's already far more complicated.
